README.md

## Pariswap Documentation

Welcome to the Pariswap documentation! This guide will provide you with all the information you need to integrate Pariswap, a dev tool built on top of the Hxro Parimutuel Protocol, into your application. Pariswap allows users to place bets or positions using any SPL token they hold, while developers can take fees in a single transaction. Pariswap also takes advantage of versioned transactions in Solana, ensuring efficiency and scalability in transaction processing.

---
![PlaceBet Popup](/placebet_screenshot.png)
---

### Table of Contents
1. Introduction
2. Getting Started
   - Installation
   - Usage Requirements
3. API Reference
   - `getPlacePositionTransaction`
   - `fetchSwappableTokenList`
4. React Hook Component
   - `usePariswapApi`
   - `PariswapProvider`
5. Usage Examples
   - Placing a Position
   - Using the React Hook Component
6. Additional Resources
   - Support
   - Examples

---

### 1. Introduction
Pariswap is a development tool that enhances the capabilities of the Hxro Parimutuel Protocol. With Pariswap, developers can create applications that enable users to engage in peer-to-peer, floating strike options. These options have a pool-based, parimutuel payoff, allowing users to deposit funds into pools representing various outcomes. When a winning outcome is determined, users receive a pro-rata distribution of the total funds in the winning pool.

Built on the Solana blockchain, Pariswap offers seamless integration through the @hxronetwork/parimutuelsdk library. This library provides a range of functions and a React hook component, simplifying the integration process for developers.

### 2. Getting Started

#### Installation
To get started with Pariswap, you need to install the required dependencies:

```bash
npm install @hxronetwork/parimutuelsdk or yarn add @hxronetwork/parimutuelsdk
```

Then, install Pariswap:

```bash
npm install pariswap or yarn add pariswap 
```

#### Usage Requirements
Before using Pariswap, ensure you have the following prerequisites:
- Knowledge of Solana blockchain and smart contract development.
- A Solana wallet and connection to a Solana network.
- Basic understanding of the Hxro Parimutuel Protocol.

### 3. API Reference
This section provides an overview of the available functions in the Pariswap API.

#### `getPlacePositionTransaction`
The `getPlacePositionTransaction` function retrieves a versioned transaction for placing a position in a parimutuel market. It can be used independently without the React hook component.

```typescript
async function getPlacePositionTransaction(
  connection: Connection,
  wallet: WalletSigner | Keypair,
  inputTokenAddress: string,
  amount: number,
  parimutuelMarket: string,
  side: PositionSideEnum,
  devUsdcATA?: string,
  devFeeBps: number = 0,
): Promise<VersionedTransaction>
```

- `connection`: The connection object for interacting with the Solana blockchain.
- `wallet`: The wallet used for signing the transaction.
- `inputTokenAddress`: The mint address of the token the user wants to use.
- `amount`: The amount in USD to be placed for the bet.
- `parimutuelMarket`: The parimutuel market where the position will be placed.
- `side`: The side of the position (e.g., `PositionSideEnum.LONG` or `PositionSideEnum.SHORT`).
- `devUsdcATA` (optional): An optional dev USDC ATA address (should be already initialized).
- `devFeeBps` (optional): Optional developer fee basis points (e.g., 50 for 0.5%, 100 for 1%, 10000 for 100%).

Returns a promise that resolves to the versioned transaction.

#### `fetchSwappableTokenList`
The `fetchSwappableTokenList` function is a helper function that fetches a list of tokens swappable with USDC. It returns only validated tokens, excluding unknown and banned tokens.

```typescript
async function fetchSwappableTokenList(): Promise<TokenData[]>
```

Returns a promise that resolves to an array of `TokenData` objects representing the swappable tokens.

### 4. React Hook Component
Pariswap provides a React hook component and a provider component to simplify the integration process.

#### `usePariswapApi`
The `usePariswapApi` hook provides a convenient way to access the Pariswap API functions within a React component.

```typescript
import { usePariswapApi } from 'pariswap';

function YourComponent() {
  const { placeBet, fetchSwappableTokenList } = usePariswapApi();
  // Your component code...
}
```

#### `PariswapProvider`
The `PariswapProvider` component wraps your main React component and provides the necessary context for using Pariswap.

```typescript
import { PariswapProvider } from 'pariswap';

function App() {
  return (
    <PariswapProvider>
      <YourContainerComponentOrMainContainer />
    </PariswapProvider>
  );
}

export default App;
```

Ensure that the `PariswapProvider` component is placed at the root level of your application.

### 5. Usage Examples

#### Placing a Position
The following example demonstrates how to use the `getPlacePositionTransaction` function to place a position in a parimutuel market.

```typescript
import {
  Connection,
  Keypair,
  PublicKey,
} from '@solana/web3.js';

import {
  ParimutuelWeb3,
  MarketPairEnum,
  PositionSideEnum,
  getMarketPubkeys,
  MAINNET_CONFIG,
} from '@hxronetwork/parimutuelsdk';

import

 { getAssociatedTokenAddressSync } from "@solana/spl-token";

const user = Keypair.generate();
const dev = Keypair.generate();
const devFeeBps = 50; // 0.50 percent
const betAmount = 1;

const config = MAINNET_CONFIG;
const desiredSeconds = 60; // 1 minute
const market = MarketPairEnum.BTCUSD;
const side = PositionSideEnum.LONG;
let rpc = process.env.RPC_URL || '';

const connection = new Connection(rpc, 'confirmed');
const parimutuelWeb3 = new ParimutuelWeb3(config, connection);
const markets = getMarketPubkeys(config, market);

const marketsByTime = markets.filter(
  (market) => market.duration === desiredSeconds
);

const parimutuels = await parimutuelWeb3.getParimutuels(marketsByTime);

const pari_markets = parimutuels.filter(
  (account) =>
    account.info.parimutuel.timeWindowStart.toNumber() > Date.now() &&
    account.info.parimutuel.timeWindowStart.toNumber() <
      Date.now() + desiredSeconds * 1000
);

const pubkey = pari_markets[0].pubkey.toString();

// Dev USDC Account (must've been initialized)
const devUsdcATA = await getAssociatedTokenAddressSync(
  new PublicKey(USDC_MINT),
  dev.publicKey
);

const placeBetTransaction = await getPlacePositionTransaction(
  connection,
  user,
  "So11111111111111111111111111111111111111112",
  betAmount,
  pubkey,
  side,
  devUsdcATA.toBase58(),
  devFeeBps
);

placeBetTransaction.sign([user]);
const txId = await connection.sendTransaction(placeBetTransaction);
console.log(`https://explorer.solana.com/tx/${txId}`);
```

#### Using the React Hook Component
The following example demonstrates how to use the `placeBet` function from the `usePariswapApi` hook to place a bet using the React hook component approach.

```typescript
import { usePariswapApi } from 'pariswap';
import { useWallet } from '@solana/wallet-adapter-react';

function YourComponent() {
  const { placeBet } = usePariswapApi();
  const wallet = useWallet();
  const amount = 1;
  const parimutuelMarket = 'parimutuel-market-address';

  const handlePlaceBet = async () => {
    // Dev USDC Account (must've been initialized)
    const devUsdcATA = await getAssociatedTokenAddressSync(
      new PublicKey(USDC_MINT),
      dev.publicKey
    );

    placeBet(
      connection,
      wallet,
      amount,
      parimutuelMarket,
      PositionSideEnum.LONG,
      devUsdcATA.toBase58(),
      50 // 0.5% developer fee
    ).then((txId) => {
      if (txId) {
        console.log(`https://explorer.solana.com/tx/${txId}`);
        notify({
          type: 'success',
          message: `Placed ${
            side === PositionSideEnum.LONG ? 'LONG' : 'SHORT'
          } Position`,
          txid: txId,
        });
      }
    });
  };

  // Your component code...

  return (
    <div>
      <button onClick={handlePlaceBet}>Place Bet</button>
    </div>
  );
}
```

### 6. Additional Resources

#### Support
If you have any questions or need assistance with Pariswap, feel free to reach out to me on [Twitter](https://www.twitter.com/femi_0x).

#### Examples
To explore more examples and usage scenarios, please refer to the [Pariswap Examples repository](https://github.com/IMEF-FEMI/pariswap/tree/main/example).

---

That's it! You should now have a good understanding of how to integrate Pariswap into your application. Happy coding!

CHANGELOG.md
# pariswap

## 1.0.34

### Patch Changes

- update doc

## 1.0.33

### Patch Changes

- update modal bg

## 1.0.3

### Patch Changes

- remove unnecesary stuff

## 1.0.2

### Patch Changes

- doc images

## 1.0.1

### Patch Changes

- 71ee3a3: doc \*\*image
- 6125dd0: updated documentation

## 1.0.0

### Major Changes

- Version 1 release: placeBet instruction generation && popup component to allow users place bet using any token of choice

## 0.3.1

### Patch Changes

- updated hooks structure

## 0.3.0

### Minor Changes

- 8cb49ca: bumping version up

## 0.2.0

### Minor Changes

- CI/CD actions setup


test/index.test.ts
import {
  USDC_MINT,
  fetchSwappableTokenList,
  getParimutuelMarketTimeLeft,
  getPlacePositionTransaction,
  getTokenValueInUSD,
} from '../dist';

import {
  ParimutuelWeb3,
  MarketPairEnum,
  getMarketPubkeys,
  PositionSideEnum,
  MAINNET_CONFIG,
} from '@hxronetwork/parimutuelsdk';
import { Connection, Keypair, PublicKey } from '@solana/web3.js';
import { getAssociatedTokenAddressSync } from 'spl-token';

require('dotenv').config();

describe('Pariswap functions', () => {
  it('fetches list of tokens that can be swapped with USDC', async () => {
    const tokenList = await fetchSwappableTokenList();
    console.log({ tokenList });

    expect(tokenList.length).toBeGreaterThan(0);
  });

  it('gets token value in USDC', async () => {
    const usdValue = await getTokenValueInUSD('SOL', 0.234);
    expect(usdValue).toBeGreaterThan(0);
  });
  it.only('generates LONG/SHORT instruction', async () => {
    const user = Keypair.generate();
    const dev = Keypair.generate();
    const devFeeBps = 50; //0.50 percent
    const betAmount = 1;

    const config = MAINNET_CONFIG;
    const desiredSeconds = 60; //1 minute
    const market = MarketPairEnum.BTCUSD;
    const side = PositionSideEnum.LONG;
    let rpc = process.env.RPC_URL || '';

    const connection = new Connection(rpc, 'confirmed');
    const parimutuelWeb3 = new ParimutuelWeb3(config, connection);
    const markets = getMarketPubkeys(config, market);

    const marketsByTime = markets.filter(
      (market) => market.duration === desiredSeconds
    );

    const parimutuels = await parimutuelWeb3.getParimutuels(marketsByTime);

    const pari_markets = parimutuels.filter(
      (account) =>
        account.info.parimutuel.timeWindowStart.toNumber() > Date.now() &&
        account.info.parimutuel.timeWindowStart.toNumber() <
          Date.now() + desiredSeconds * 1000
    );

    const pubkey = pari_markets[0].pubkey.toString();

    //dev usdc Account (must've been initialized)
    const devUsdcATA = await getAssociatedTokenAddressSync(
      new PublicKey(USDC_MINT),
      dev.publicKey
    );
    const placeBetTransaction = await getPlacePositionTransaction(
      connection,
      user,
      'GzpRsvnKXKz586kRLkjdppR4dUCFwHa2qaszKkPUQx6g',
      betAmount,
      pubkey,
      side,
      devUsdcATA.toBase58(),
      devFeeBps
    );

    placeBetTransaction.sign([user]);
    const txId = await connection.sendTransaction(placeBetTransaction);
    console.log(`https://explorer.solana.com/tx/${txId}`);
  });

  it('gets time left for parimutuel market', async () => {
    const config = MAINNET_CONFIG;
    const desiredSeconds = 60; //1 minute
    const market = MarketPairEnum.BTCUSD;
    const side = PositionSideEnum.LONG;
    let rpc = process.env.RPC_URL || '';

    const connection = new Connection(rpc, 'confirmed');
    const parimutuelWeb3 = new ParimutuelWeb3(config, connection);
    const markets = getMarketPubkeys(config, market);

    const marketsByTime = markets.filter(
      (market) => market.duration === desiredSeconds
    );

    const parimutuels = await parimutuelWeb3.getParimutuels(marketsByTime);

    const pari_markets = parimutuels.filter(
      (account) =>
        account.info.parimutuel.timeWindowStart.toNumber() > Date.now() &&
        account.info.parimutuel.timeWindowStart.toNumber() <
          Date.now() + desiredSeconds * 1000
    );

    const pubkey = pari_markets[0].pubkey.toString();
    const timeLeft = await getParimutuelMarketTimeLeft(
      connection,
      new PublicKey(pubkey)
    );
    expect(timeLeft).toBeTruthy();
  });
});


src/utils/index.ts
import { Connection, PublicKey, TransactionInstruction } from '@solana/web3.js';

import { ParimutuelWeb3, MAINNET_CONFIG } from '@hxronetwork/parimutuelsdk';
import {
  getAssociatedTokenAddressSync,
  getAccount,
  createAssociatedTokenAccountInstruction,
  TOKEN_PROGRAM_ID,
} from 'spl-token';
import { Route, TokenData } from '../types';

export const USDC_MINT = 'EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v';
export const SWAP_FEE_BPS = 50;
export const SWAP_FEE_WALLET = '5qxvvD5fJJK2xPBmifNV1cmbN6vMrncjTinmFWU1eRGs';

/**
 * Fetches a list of tokens swappable with USDC. (returns only validated tokens -- No unknown and banned tokens).
 *
 * @returns {Promise<TokenData[]>} - A promise that resolves to an array of TokenData objects representing the swappable tokens.
 */
export async function fetchSwappableTokenList(): Promise<TokenData[]> {
  // Fetch token list
  const response = await fetch('https://token.jup.ag/strict');
  const tokenList: TokenData[] = (await response.json()) as Array<any>;

  // Retrieve indexed routed map
  const indexedRouteMapResponse = await fetch(
    'https://quote-api.jup.ag/v4/indexed-route-map'
  );
  const indexedRouteMap = await indexedRouteMapResponse.json();

  const getMint = (index): string => indexedRouteMap['mintKeys'][index];

  // Generate route map by replacing indexes with mint addresses
  const generatedRouteMap = {};

  Object.keys(indexedRouteMap['indexedRouteMap']).forEach((key) => {
    generatedRouteMap[getMint(key)] = indexedRouteMap['indexedRouteMap'][
      key
    ].map((index: string | number) => getMint(index));
  });

  // const tokens = getPossiblePairsTokenInfo({ tokens: tokenList, routeMap: generatedRouteMap, inputTokenAddress: USDC_MINT })

  // console.log(tokens.l);

  // return Object.values(tokens).filter((token) => {
  //   if (token) return true
  //   else return false
  // });

  const swappableOutputForUSDC = generatedRouteMap[USDC_MINT];
  const getIndexInUSDCSwappableOutput = (mint: string) =>
    swappableOutputForUSDC.indexOf(mint);

  const filteredTokenList: TokenData[] = tokenList.filter(
    (token: TokenData) => getIndexInUSDCSwappableOutput(token.address) !== -1
  );

  return filteredTokenList;
}

// A helper function to help us find which output pair is possible
const getPossiblePairsTokenInfo = ({
  tokens,
  routeMap,
  inputTokenAddress,
}: {
  tokens: TokenData[];
  routeMap: { [key: string]: string[] };
  inputTokenAddress?: string;
}) => {
  try {
    const possiblePairs = routeMap[inputTokenAddress];
    var possiblePairsTokenInfo: { [key: string]: TokenData | undefined } = {};
    possiblePairs.forEach((address) => {
      possiblePairsTokenInfo[address] = tokens.find((t) => {
        return t.address == address;
      });
    });

    return possiblePairsTokenInfo;
  } catch (error) {
    throw error;
  }
};
export async function getExactOutQuote(amount: number, inputMint: string) {
  const { data } = await (
    await fetch(
      `https://quote-api.jup.ag/v4/quote?inputMint=${inputMint}&outputMint=${USDC_MINT}&amount=${amount}&slippageBps=50&feeBps=${SWAP_FEE_BPS}&swapMode=ExactOut`
    )
  ).json();
  return data;
}

export async function getExactInQuote(amount: number, inputMint: string) {
  const { data } = await (
    await fetch(
      `https://quote-api.jup.ag/v4/quote?inputMint=${inputMint}&outputMint=${USDC_MINT}&amount=${amount}&slippageBps=50&feeBps=${SWAP_FEE_BPS}&swapMode=ExactIn`
    )
  ).json();
  return data;
}

export async function getTokenPrice(inputMint: string) {
  const { data } = await (
    await fetch(
      `https://price.jup.ag/v4/price?ids=${inputMint}&vsToken=${USDC_MINT}`
    )
  ).json();

  if (!data[inputMint]) {
    return 0;
  }
  return Number(data[inputMint].price);
}

export async function getTokenValueInUSD(token: string, amount: number) {
  const { data } = await (
    await fetch(`https://price.jup.ag/v4/price?ids=${token}`)
  ).json();

  if (!data || Object.keys(data).length === 0) return 0;

  return amount * Number(data[token].price);
}

export async function getSwapTransaction(
  routes: Route[],
  wallet: string,
  swapType: string
) {
  let feeAccountATA: PublicKey;
  if (swapType === 'ExactIn') {
    feeAccountATA = await getAssociatedTokenAddressSync(
      new PublicKey(USDC_MINT),
      new PublicKey(SWAP_FEE_WALLET)
    );
  } else {
    feeAccountATA = await getAssociatedTokenAddressSync(
      new PublicKey(routes[0].marketInfos[0].inputMint),
      new PublicKey(SWAP_FEE_WALLET)
    );
  }

  // get serialized transactions for the swap
  const transactions = await (
    await fetch('https://quote-api.jup.ag/v4/swap', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        // route from /quote api
        route: routes[0],
        // user public key to be used for the swap
        userPublicKey: wallet,
        // auto wrap and unwrap SOL. default is true
        wrapUnwrapSOL: swapType === 'ExactOut',
        // feeAccount is optional. Use if you want to charge a fee.  feeBps must have been passed in /quote API.
        // This is the ATA account for the output token where the fee will be sent to. If you are swapping from SOL->USDC then this would be the USDC ATA you want to collect the fee.
        feeAccount: feeAccountATA,
      }),
    })
  ).json();

  const { swapTransaction } = transactions;

  return swapTransaction;
}

export async function getParimutuelMarketTimeLeft(
  connection: Connection,
  parimutuelMarket: PublicKey
) {
  const parimutuelWeb3 = new ParimutuelWeb3(MAINNET_CONFIG, connection);
  const pariMarket = await parimutuelWeb3.getParimutuel(parimutuelMarket);
  const timeWindowStart = pariMarket.info.parimutuel.timeWindowStart.toNumber();
  const currentTime = new Date().getTime();
  return timeWindowStart - currentTime;
}

export async function getPlatformFeeAccounts(
  connection: Connection,
  feeAccountOwner: PublicKey
) {
  const _userTokens = await connection.getParsedTokenAccountsByOwner(
    feeAccountOwner,
    { programId: TOKEN_PROGRAM_ID }
  );

  const accounts = _userTokens.value.map((account) => {
    const publicKey = account.pubkey;
    const accountInfo = account.account.data.parsed.info;
    const mint = accountInfo.mint;
    const balance = accountInfo.tokenAmount.amount;

    return { publicKey, mint, balance };
  });

  const feeAccounts = accounts.filter((account) => account.mint === USDC_MINT);
  return feeAccounts;
}

export async function getCreateAccountInstruction(
  connection: Connection,
  payer: PublicKey,
  mint: PublicKey,
  owner: PublicKey,
  ATA: PublicKey
): Promise<TransactionInstruction | undefined> {
  try {
    // return null if account exists
    await getAccount(connection, ATA);
    return undefined;
  } catch (error: unknown) {
    return createAssociatedTokenAccountInstruction(payer, ATA, owner, mint);
  }
}


example/src/utils/helpers.ts

import { getAccount, getAssociatedTokenAddress } from "@solana/spl-token"
import { Connection, PublicKey } from "@solana/web3.js"


export const getBalance = async (connection: Connection, owner: PublicKey, isNative = false, tokenMint: PublicKey) => {
    if (isNative) {
        const userBalance = await connection.getBalance(owner)
        return toUIAmount(userBalance, 9)
    } else {
        const userATA = await getATA(connection, tokenMint, owner,)
        if (userATA) {
            const userBalance = await connection.getTokenAccountBalance(userATA.address)
            return Number(userBalance.value.uiAmount);
        }
        return 0;
    }
}


export const getATA = async (connection, tokenMint, owner) => {
    const userATA = await getAssociatedTokenAddress(
        tokenMint,
        owner,
        true
    )

    let account;
    try {
        account = await getAccount(connection, userATA,);
    } catch (error) {

    }
    // if getAccount throws error, send null
    //meaning user does not have that token
    return account;
}
export const fromUIAmount = (amount, decimals) => Number(amount) * 10 ** decimals
export const toUIAmount = (amount, decimals) => Number(amount) / (10 ** decimals)

src/instructions/placePositions.ts
import {
  MAINNET_CONFIG,
  ParimutuelWeb3,
  PositionSideEnum,
  WalletSigner,
} from '@hxronetwork/parimutuelsdk';
import {
  AddressLookupTableAccount,
  Connection,
  Keypair,
  PublicKey,
  TransactionInstruction,
  TransactionMessage,
  VersionedTransaction,
} from '@solana/web3.js';
import {
  SWAP_FEE_WALLET,
  USDC_MINT,
  getCreateAccountInstruction,
  getExactInQuote,
  getExactOutQuote,
  getSwapTransaction,
  getTokenPrice,
} from '../utils';
import {
  createTransferInstruction,
  getAssociatedTokenAddressSync,
} from 'spl-token';
import { TokenData } from '../types';

/**
 * Retrieves a versioned transaction for placing a position in a parimutuel market.
 *
 * @param {Connection} connection - The connection object for interacting with Solana blockchain.
 * @param {WalletSigner | Keypair} wallet - The wallet used for signing the transaction.
 * @param {string} inputTokenAddress - The mint address of the token the user wants to use.
 * @param {number} amount - The amount in USD to be placed for the Bet.
 * @param {string} parimutuelMarket - The parimutuel market where the position will be placed.
 * @param {PositionSideEnum} side - The side of the position (e.g., PositionSideEnum.LONG orPositionSideEnum.SHORT).
 * @param {string} [devUsdcATA] - An Optional dev USDC ATA address (should be already initialized).
 * @param {number} [devFeeBps=0] - Optional developer fee basis points (e.g. 50 for 0.5%, 100 for 1% ... 10000 for 100%).
 * @returns {Promise<VersionedTransaction>} - A promise that resolves to the versioned transaction.
 */
export async function getPlacePositionTransaction(
  connection: Connection,
  wallet: WalletSigner | Keypair,
  inputTokenAddress: string,
  amount: number,
  parimutuelMarket: string,
  side: PositionSideEnum,
  devUsdcATA?: string,
  devFeeBps: number = 0
): Promise<VersionedTransaction> {
  amount = amount * 1e6;

  const userUsdcATA = await getAssociatedTokenAddressSync(
    new PublicKey(USDC_MINT),
    wallet.publicKey
  );
  if (inputTokenAddress === USDC_MINT) {
    await ensureAccountHasBeenInitialized(connection, userUsdcATA);

    const balance = await connection.getTokenAccountBalance(userUsdcATA);
    if (Number(balance.value.amount) < amount) {
      throw new Error('Insufficient funds');
    }

    return await getPlacePositionTransactionWithoutSwap(
      connection,
      wallet,
      amount,
      parimutuelMarket,
      side,
      devUsdcATA,
      devFeeBps
    );
  } else {
    return await getPlacePositionTransactionWithSwap(
      connection,
      wallet,
      inputTokenAddress,
      amount,
      parimutuelMarket,
      side,
      devUsdcATA,
      devFeeBps
    );
  }
}
async function getPlacePositionTransactionWithoutSwap(
  connection: Connection,
  wallet: WalletSigner | Keypair,
  amount: number,
  parimutuelMarket: string,
  side: PositionSideEnum,
  devUsdcATA?: string,
  devFeeBps: number = 0
): Promise<VersionedTransaction> {
  const parimutuelWeb3 = new ParimutuelWeb3(MAINNET_CONFIG, connection);

  const devFee = amount * (devFeeBps / 10_000);
  const betAmount = amount - devFee;

  const instructions: TransactionInstruction[] = [];

  //HANDLE DEV FEE TRANSFER
  if (devUsdcATA && devFeeBps > 0) {
    const userUsdcATA = await getAssociatedTokenAddressSync(
      new PublicKey(USDC_MINT),
      wallet.publicKey
    );

    const transferInstruction = createTransferInstruction(
      userUsdcATA,
      new PublicKey(devUsdcATA),
      wallet.publicKey,
      devFee,
      []
    );
    //transfer fee to dev wallet instruction
    instructions.push(transferInstruction);
  }

  let placePositionIx = await parimutuelWeb3.getPlacePositionInstruction(
    wallet as WalletSigner,
    new PublicKey(parimutuelMarket),
    betAmount,
    side,
    Date.now()
  );
  instructions.push(...placePositionIx);

  //convert to VersionedTransaction
  //
  // Fetch Latest Blockhash
  let latestBlockhash = await connection.getLatestBlockhash('confirmed');

  //Generate Transaction Message
  const messageV0 = new TransactionMessage({
    payerKey: wallet.publicKey,
    recentBlockhash: latestBlockhash.blockhash,
    instructions,
  }).compileToV0Message();

  return new VersionedTransaction(messageV0);
}

async function getPlacePositionTransactionWithSwap(
  connection: Connection,
  wallet: WalletSigner | Keypair,
  inputTokenAddress: string,
  amount: number,
  parimutuelMarket: string,
  side: PositionSideEnum,
  devUsdcATA?: string,
  devFeeBps: number = 0
): Promise<VersionedTransaction> {
  const parimutuelWeb3 = new ParimutuelWeb3(MAINNET_CONFIG, connection);

  //get quote routes
  const exactOutRoutes = await getExactOutQuote(amount, inputTokenAddress);
  if (exactOutRoutes.length > 0) {
    return getPlacePositionTransactionWithSwapExactOut(
      connection,
      parimutuelWeb3,
      wallet,
      exactOutRoutes,
      amount,
      parimutuelMarket,
      side,
      devUsdcATA,
      devFeeBps
    );
  } else {
    return getPlacePositionTransactionWithSwapExactIn(
      connection,
      parimutuelWeb3,
      wallet,
      inputTokenAddress,
      amount,
      parimutuelMarket,
      side,
      devUsdcATA,
      devFeeBps
    );
  }
}

async function getPlacePositionTransactionWithSwapExactIn(
  connection: Connection,
  parimutuelWeb3: ParimutuelWeb3,
  wallet: WalletSigner | Keypair,
  inputTokenAddress: string,
  amount: number,
  parimutuelMarket: string,
  side: PositionSideEnum,
  devUsdcATA?: string,
  devFeeBps: number = 0
) {
  const devFee = amount * (devFeeBps / 10_000);
  const betAmount = amount - devFee;

  //price
  const tokenPrice = await getTokenPrice(inputTokenAddress);
  const amountOfInput = amount / tokenPrice;
  const padding = (amount / tokenPrice) * 0.05;
  // console.log(Math.ceil(amountOfInput + padding));
  const routes = await getExactInQuote(
    Math.ceil(amountOfInput + padding),
    inputTokenAddress
  );

  const swapTransaction = await getSwapTransaction(
    routes,
    wallet.publicKey.toBase58(),
    'ExactIn'
  );
  const swapTransactionBuf = Buffer.from(swapTransaction, 'base64');
  var transaction = VersionedTransaction.deserialize(swapTransactionBuf);

  // get address lookup table accounts
  const addressLookupTableAccounts = await Promise.all(
    transaction.message.addressTableLookups.map(async (lookup) => {
      return new AddressLookupTableAccount({
        key: lookup.accountKey,
        state: AddressLookupTableAccount.deserialize(
          await connection.getAccountInfo(lookup.accountKey).then((res) => {
            return res.data;
          })
        ),
      });
    })
  );
  // decompile transaction message and add transfer instruction
  var message = TransactionMessage.decompile(transaction.message, {
    addressLookupTableAccounts: addressLookupTableAccounts,
  });

  //HANDLE DEV FEE TRANSFER
  if (devUsdcATA && devFeeBps > 0) {
    const userUsdcATA = await getAssociatedTokenAddressSync(
      new PublicKey(USDC_MINT),
      wallet.publicKey
    );

    const transferInstruction = createTransferInstruction(
      userUsdcATA,
      new PublicKey(devUsdcATA),
      wallet.publicKey,
      devFee,
      []
    );

    message.instructions.push(transferInstruction);
  }

  let placePositionIx = await parimutuelWeb3.getPlacePositionInstruction(
    wallet as WalletSigner,
    new PublicKey(parimutuelMarket),
    betAmount,
    side,
    Date.now()
  );
  message.instructions.push(...placePositionIx);
  transaction.message = message.compileToV0Message(addressLookupTableAccounts);

  return transaction;
}
async function getPlacePositionTransactionWithSwapExactOut(
  connection: Connection,
  parimutuelWeb3: ParimutuelWeb3,
  wallet: WalletSigner | Keypair,
  routes: any[],
  amount: number,
  parimutuelMarket: string,
  side: PositionSideEnum,
  devUsdcATA?: string,
  devFeeBps: number = 0
) {
  const devFee = amount * (devFeeBps / 10_000);
  const betAmount = amount - devFee;

  //get swap instruction
  const swapTransaction = await getSwapTransaction(
    routes,
    wallet.publicKey.toBase58(),
    'ExactOut'
  );
  const swapTransactionBuf = Buffer.from(swapTransaction, 'base64');
  var transaction = VersionedTransaction.deserialize(swapTransactionBuf);

  // get address lookup table accounts
  const addressLookupTableAccounts = await Promise.all(
    transaction.message.addressTableLookups.map(async (lookup) => {
      return new AddressLookupTableAccount({
        key: lookup.accountKey,
        state: AddressLookupTableAccount.deserialize(
          await connection.getAccountInfo(lookup.accountKey).then((res) => {
            return res.data;
          })
        ),
      });
    })
  );
  // decompile transaction message and add transfer instruction
  var message = TransactionMessage.decompile(transaction.message, {
    addressLookupTableAccounts: addressLookupTableAccounts,
  });

  //add instruction to create fee ATA if none exists
  const feeATA = getAssociatedTokenAddressSync(
    new PublicKey(routes[0].marketInfos[0].inputMint),
    new PublicKey(SWAP_FEE_WALLET)
  );
  const feeATACreateInstruction = await getCreateAccountInstruction(
    connection,
    wallet.publicKey,
    new PublicKey(routes[0].marketInfos[0].inputMint),
    new PublicKey(SWAP_FEE_WALLET),
    feeATA
  );
  if (feeATACreateInstruction) {
    message.instructions.unshift(feeATACreateInstruction);
  }
  //HANDLE DEV FEE TRANSFER
  if (devUsdcATA && devFeeBps > 0) {
    const userUsdcATA = await getAssociatedTokenAddressSync(
      new PublicKey(USDC_MINT),
      wallet.publicKey
    );

    const transferInstruction = createTransferInstruction(
      userUsdcATA,
      new PublicKey(devUsdcATA),
      wallet.publicKey,
      devFee,
      []
    );
    //transfer fee to dev wallet instruction
    message.instructions.push(transferInstruction);
  }

  let placePositionIx = await parimutuelWeb3.getPlacePositionInstruction(
    wallet as WalletSigner,
    new PublicKey(parimutuelMarket),
    betAmount,
    side,
    Date.now()
  );
  message.instructions.push(...placePositionIx);
  transaction.message = message.compileToV0Message(addressLookupTableAccounts);

  return transaction;
}

async function ensureAccountHasBeenInitialized(
  connection: Connection,
  account: PublicKey
) {
  try {
    return await connection.getAccountInfo(account);
  } catch (error) {
    throw new Error('Account Has not been initialized');
  }
}
